src/main/java/io/vertx/core/eventbus/impl/EventBusImpl.java

```diff
@@ -736,17 +750,26 @@ private void checkStarted() {
     }
   }
 
-  private static class HandlerHolder<T> {
+  private class HandlerHolder<T> {
     final ContextImpl context;
-    final Handler<Message<T>> handler;
+    final HandlerRegistration<T> handler;
     final boolean replyHandler;
     final boolean localOnly;
     final long timeoutID;
     boolean removed;
 
     // We use a synchronized block to protect removed as it can be unregistered from a different thread
-    synchronized void setRemoved() {
-      removed = true;
+    void setRemoved() {
```

Comment:
the synchronization is done in a synchronized block in the method to have the metrics.handlerUnregistered callback to be done out of this block.